File: array/sliding_window/Solution424.java
public class Solution424 {
    public int characterReplacement(String s, int k) {
        if (s == null){
            return 0;
        }
        int[] count = new int[26];
        int start = 0;     // 因为找到最大值即可，窗口并不需要收窄。并不需要维护一个在窗口内恒成立的条件。
        int maxCount = 0;  // maxCount并不需要是“窗口内”字母的最大出现次数 
        for (int end = 0; end < s.length(); end++){
            int key = s.charAt(end) - 'A';
            count[key]++;
            maxCount = Math.max(maxCount, count[key]);
            if (end - start + 1 > maxCount + k){
                count[s.charAt(start) - 'A']--;
                start++;
            }
        }
        return s.length() - start;
    }
}

File: array/sliding_window/Solution424.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    const count = new Array(26).fill(0);
    let maxCount = 0;
    let start = 0;
    const length = s.length;
    for (let end = 0; end < length; end++){
        let idx = s[end].charCodeAt() - 'A'.charCodeAt();
        count[idx]++;
        maxCount = Math.max(maxCount, count[idx]);
        if (end - start + 1 > maxCount + k){
            count[s[start].charCodeAt() - 'A'.charCodeAt()]--;
            start++;
        }
    }
    return length - start;
};

let ans = characterReplacement("AABABBA", 1);
console.log(ans);